bpfman/src/lib.rs

```diff
@@ -109,1139 +98,1009 @@ pub(crate) fn get_db_config() -> SledConfig {
+            );
+            sleep(Duration::from_millis(database_config.millisec_delay)).await;
         }
     }
+    Err(BpfmanError::DatabaseLockError)
+}
+
+// Make sure to call init_image_manger if the command requires interaction with
+// an OCI based container registry. It should ONLY be used where needed, to
+// explicitly control when bpfman blocks for network calls to both sigstore's
+// cosign tuf registries and container registries.
+pub(crate) async fn init_image_manager() -> ImageManager {
+    let config = open_config_file();
+    ImageManager::new(config.signing().as_ref().map_or(true, |s| s.allow_unsigned))
+        .await
+        .expect("failed to initialize image manager")
+}
 
-    pub(crate) async fn add_multi_attach_program(
-        &mut self,
-        program: &mut Program,
-    ) -> Result<u32, BpfmanError> {
-        debug!("BpfManager::add_multi_attach_program()");
-        let name = &program.get_data().get_name()?;
-
-        // This load is just to verify the BPF Function Name is valid.
-        // The actual load is performed in the XDP or TC logic.
-        // don't pin maps here.
-        let mut ext_loader = BpfLoader::new()
-            .allow_unsupported_maps()
-            .extension(name)
-            .load(&program.get_data().get_program_bytes()?)?;
-
-        match ext_loader.program_mut(name) {
-            Some(_) => Ok(()),
-            None => Err(BpfmanError::BpfFunctionNameNotValid(name.to_owned())),
-        }?;
-
-        let did = program
-            .dispatcher_id()?
-            .ok_or(BpfmanError::DispatcherNotRequired)?;
-
-        let next_available_id = self.attached_programs(&did);
-        if next_available_id >= 10 {
-            return Err(BpfmanError::TooManyPrograms);
+fn get_dispatcher(id: &DispatcherId, root_db: &Db) -> Option<Dispatcher> {
+    let tree_name_prefix = match id {
+        DispatcherId::Xdp(DispatcherInfo(if_index, _)) => {
+            format!("{}_{}", XDP_DISPATCHER_PREFIX, if_index)
+        }
+        DispatcherId::Tc(DispatcherInfo(if_index, Some(direction))) => {
+            format!("{}_{}_{}", TC_DISPATCHER_PREFIX, if_index, direction)
         }
+        _ => {
+            return None;
+        }
+    };
 
-        debug!("next_available_id={next_available_id}");
+    root_db
+        .tree_names()
+        .into_iter()
+        .find(|p| bytes_to_string(p).contains(&tree_name_prefix))
+        .map(|p| {
+            let tree = root_db.open_tree(p).expect("unable to open database tree");
+            Dispatcher::new_from_db(tree)
+        })
+}
 
-        let program_type = program.kind();
-        let if_index = program.if_index()?;
-        let if_name = program.if_name().unwrap().to_string();
-        let direction = program.direction()?;
+/// Returns the number of extension programs currently attached to the dispatcher that
+/// would be used to attach the provided [`Program`].
+fn attached_programs(did: &DispatcherId, root_db: &Db) -> usize {
```

Comment:
I know you didn't change this, but I think it would be more clear to name this function num_attached_programs.

Reply:
Done!